### PR TITLE
[batch] improve clarity

### PR DESCRIPTION
**Pull request by cseed**

### CODE REVIEW

**Comment by cseed** [comment]
This doesn't necessarily fix the problem, but I think they are good changes in the direction of what we know.  It passed test_{batch, pipeline} on the first try, so that's a good sign.  What I did:

1. Lock down Job state transitions.  Now only set_state and _mark_job_task_complete change _state, and they log identically.  Explicitly enumerate the valid state transitions are check them in each function.  Slightly clarified the transitions around Pending.  Now Pending can only go to Ready.

2. If a state update fails (the Python object is stale), throw JobStateWriteFailure.  If we have a stale picture, we clearly don't want to be doing anything else.

3. Handle a few more cases in update_job_with_pod: a pod without a job or a job that shouldn't have one, and a cancellable pod that hasn't been cancelled yet.

File: batch/batch/batch.py [hunk 1]
```diff
@@ -541,9 +542,10 @@ def is_successful(self):
         return self._state == 'Success'
 
     async def mark_unscheduled(self):
-        await self._delete_pod()
-        if not self._cancelled or self.always_run:
+        if self._status == 'Running':
             await self.set_state('Ready')
+
+        if self._state == 'Ready' and (not self._cancelled or self.always_run):
             await self._create_pod()
```
Comment (danking) [inline batch/batch/batch.py]: If you try this without deleting the pod, it will fail, no?

File: batch/batch/globals.py [hunk 2]
```diff
@@ -1,2 +1,14 @@
 
+states = {'Pending', 'Ready', 'Running', 'Cancelled', 'Error', 'Failed', 'Success'}
+
 complete_states = ('Cancelled', 'Error', 'Failed', 'Success')
+
+valid_state_transitions = {
+    'Pending': {'Ready'},
+    'Ready': {'Running', 'Error', 'Cancelled'},
```
Comment (danking) [inline batch/batch/globals.py]: These first two could jump to Success in rare cases that a second batch ran a pod to completion but died before reaping it. Then this batch learns of the pods existence.

File: batch/batch/batch.py (hunk 1 above)
Comment (cseed) [reply batch/batch/batch.py]: Ah, I see.  I thought mark_unscheduled meant "there is no pod," but it really means "schedule or reschedule".  Will fix.

File: batch/batch/globals.py (hunk 2 above)
Comment (cseed) [reply batch/batch/globals.py]: I ignore pods if I'm pending:

```
    if job and job._state == 'Pending':
        if pod:
            log.error('job {job.full_id} has pod {pod.metadata.name}, ignoring')
        return
```

If another batch started it, it means my data is stale, and I should just retry later.

I think I'm inclined to get rid of the distinct between Ready and Running.  Basically, if we have a pod in hand (or None), we know it is running (or ready), and this state seems redundant.  Will have the changes ready shortly.

**Comment by cseed** [comment]
I added Running -> Running because that happens when we finish a job task.

**Pull request merged**